### PR TITLE
audit log

### LINKED ISSUE #2057: Viewing an application should be logged as an event
Currently, only actions are logged as events on an application. However, it would be necessary to log also who has opened the application just to view it (handlers, specifically) in order to maintain a good audit log.

### PR DESCRIPTION
**Pull request by opqdonut**
for #2057

# Definition of Done / Review checklist

## Reviewability
- [x] link to issue

## API
- [x] API is documented and shows up in Swagger UI
- [x] API is backwards compatible or completely new

## Documentation
- [x] update changelog if necessary
- [ ] documentation _at least_ for config options (i.e. docs folder)

## Testing
- [x] valuable features are integration / browser / acceptance tested automatically

## Follow-up
- [x] no critical TODOs left to implement

### CODE REVIEW

File: resources/migrations/20200427092452-audit-log.down.sql [hunk 1]
```diff
@@ -0,0 +1 @@
+DROP TABLE audit_log CASCADE;
```
Comment (Macroz) [inline resources/migrations/20200427092452-audit-log.down.sql:1]: I wonder if we should just stop writing these (potentially data lossing) down migrations especially for high value data.

File: src/clj/rems/api.clj [hunk 2]
```diff
@@ -90,6 +95,23 @@
 (defn- read-only? [request]
   (not (contains? #{:put :post} (:request-method request))))
 
+;; This should be run outside transaction-middleware since we want to
+;; write even on GET queries. We're only running one insert statement
+;; so we don't need a separte transaction for logging.
```
Comment (Macroz) [inline src/clj/rems/api.clj:100]: Typo

File: src/clj/rems/api.clj [hunk 3]
```diff
@@ -101,7 +123,10 @@
   request)
 
 (def api-routes
-  (api
+  ;; we wrap audit-log-middleware outside compojure-api since we want
+  ;; to see the status codes produced by our exception handlers
+  (audit-log-middleware
```
Comment (Macroz) [inline src/clj/rems/api.clj:128]: I wonder if it should be a config option to audit log or not?

Comment (Macroz) [reply src/clj/rems/api.clj:128]: Possibly need to implement some configurable deleting / vacuuming of old data.

File: test/clj/rems/api/test_api.clj [hunk 4]
```diff
@@ -1,6 +1,7 @@
 (ns ^:integration rems.api.test-api
   (:require [clojure.test :refer :all]
             [rems.api.testing :refer :all]
+            [rems.db.core :as db]
```
Comment (Macroz) [inline test/clj/rems/api/test_api.clj:4]: Why is this added?

File: test/clj/rems/api/test_audit_log.clj [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+                                              (header "Cookie" cookie)
+                                              (header "x-csrf-token" csrf)
+                                              (json-body {:id 9999999 :archived true})
+                                              handler))))))))
+    (testing "user can't get log"
+      (is (response-is-forbidden? (-> (request :get "/api/audit-log")
+                                      (authenticate "42" "alice")
+                                      handler))))
+    (testing "reporter can get log"
+      (is (= [{:userid nil :apikey nil :method "get" :path "/api/unknown" :status "404"}
```
Comment (Macroz) [inline test/clj/rems/api/test_audit_log.clj:70]: One of the acceptance test cases is that can the reporter see who has viewed a specific application? I don't think this is covered here yet. Moreover, isn't it a requirement that someone can figure out who has seen someone's data? How would one figure out which applications a person is in, and then filter from this log all the people who have seen such an application? That could be its own API I guess.

File: resources/migrations/20200427092452-audit-log.down.sql (hunk 1 above)
Comment (opqdonut) [reply resources/migrations/20200427092452-audit-log.down.sql:1]: That's a good question. However we use migratus reset to empty the db in some tests, so perhaps a middle ground would be to rename the table here?

File: src/clj/rems/api.clj (hunk 3 above)
Comment (opqdonut) [reply src/clj/rems/api.clj:128]: Cleaning data is luckily easy to do on the DB level. Let's wait for what the users need. I'm a bit hesitant to add auto-cleaning to REMS, as e.g. a system clock glitch might wipe the whole audit log.

File: test/clj/rems/api/test_api.clj (hunk 4 above)
Comment (opqdonut) [reply test/clj/rems/api/test_api.clj:4]: Leftover from earlier commit, removed.

File: test/clj/rems/api/test_audit_log.clj (hunk 5 above)
Comment (opqdonut) [reply test/clj/rems/api/test_audit_log.clj:70]: Yeah some further work on querying might be needed. Let's discuss and make a new issue? I think this feature makes sense to merge already in this state.

File: resources/migrations/20200427092452-audit-log.down.sql (hunk 1 above)
Comment (Macroz) [reply resources/migrations/20200427092452-audit-log.down.sql:1]: I guess for tests it may make sense to have some other reset function that is faster i.e. truly start from an empty DB or db dump. I think this can be left here for now.

File: src/clj/rems/api.clj (hunk 3 above)
Comment (Macroz) [reply src/clj/rems/api.clj:128]: It could be provided as an ops script sometime later.

File: test/clj/rems/api/test_audit_log.clj (hunk 5 above)
Comment (Macroz) [reply test/clj/rems/api/test_audit_log.clj:70]: Ok as long as the concern is in some task on the board.

Comment (opqdonut) [reply test/clj/rems/api/test_audit_log.clj:81]: I created #2147, and added filtering by application to this PR

File: test/clj/rems/api/test_audit_log.clj [hunk 6]
```diff
@@ -29,8 +31,15 @@
             (testing "> authorized"
               (is (= 200 (:status (-> (request :get "/api/users/active")
                                       (authenticate "42" "owner")
+                                      handler)))))
+            (reset! time-a (time/now))
+            (testing "> application"
+              (is (= 200 (:status (-> (request :get (str "/api/applications/" app-id))
```
Comment (Macroz) [inline test/clj/rems/api/test_audit_log.clj:37]: Should we prefer `api-call` and `api-response` for these in the future? We also have the `(true? :success)` and `(= {:success true}` variants. Could add some idea task to unify these a bit.

Comment (opqdonut) [reply test/clj/rems/api/test_audit_log.clj:37]: I think here it's more natural to use the explicit forms since some calls lack `authenticate`. I'll clean up a bit otherwise.
